factor/grace/classes/factor.php

```diff
@@ -0,0 +1,158 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Grace period factor class.
+ *
+ * @package     tool_mfa
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_grace;
+
+defined('MOODLE_INTERNAL') || die();
+
+use tool_mfa\local\factor\object_factor_base;
+
+class factor extends object_factor_base {
+
+    /**
+     * Grace Factor implementation.
+     * This factor needs no user setup, return true.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_user_factor($data) {
+        return true;
+    }
+
+    /**
+     * Grace Factor implementation.
+     * This factor is a singleton, return single instance.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors() {
+        global $DB, $USER;
+
+        if (!$DB->record_exists('factor_grace', array('userid' => $USER->id))) {
```

Comment:
There is nothing bad about this but I I think this method I think could be streamlined and be a little more idiomatic moodle code.

insert_record will return the id of the new record, so they way I'd do this is:

1) get_records with s so we get an array (not record_exists), if it exists then we have the full record and we're done, we can just return the array as is and exit early
2) if it is null then build the new record
3) then insert it and set id the id, now we have the same as above with no need for an extra db call, wrap in array and return
4) I would return the record as is with no translation, ie just same all those default values into the record

I'm still a bit weirded out by each factor instance having their own tables which are all the same, but we'll worry about that in #50 

The more I thin about it the more we definitely want a unified table for all factors and for all the factors, even the no input one, to still create and update real records because we need to report on how much and how recently each factor has been used.

Reply:
A little bit of translation is still required to add 'revoked', however this will be addressed in #50